R/addMetrics.R

```diff
@@ -0,0 +1,101 @@
+#' @param thresholds \code{Named list}. A list of thresholds for each metric. 
+#'   For example, \code{list(CRT = 2)} specifies the threshold for CRT 
+#'   calculation. (Default: \code{list(CRT = 2)})
+#' @param split.by \code{Character scalar}. Specifies how to split the data 
+#'   for per-group or per-subject calculation. For example, \code{"subject"}. 
+#'   If \code{NULL}, no splitting will occur. (Default: \code{NULL})
+#' @param ... Additional arguments to pass to individual metric functions. 
+#' @return A \code{\link{SummarizedExperiment}} object with added metadata 
+#'   containing the calculated metrics.
+#' 
+NULL
+
+#' @export
+setGeneric("addMetrics", function(
+    x, assay.type = "counts", 
+    metrics = c("CRT"), thresholds = list(CRT = 2), split.by = NULL, ...) {
+    standardGeneric("addMetrics")
+})
+
+#' @rdname addMetrics
+#' @export
+setMethod("addMetrics", signature = c(x = "SummarizedExperiment"),
+    function(
+        x, assay.type = "counts", 
+        metrics = c("CRT"), thresholds = list(CRT = 2), split.by = NULL, ...){
+
+        ############################## Input checks ############################
+        # Validate 'metrics'
+        if (!.is_non_empty_character(metrics)) {
+            stop("'metrics' should be a non-empty character vector.", 
+                 call. = FALSE)
+        }
+        # Check if 'split.by' exists in colData, otherwise do not split
+        if (!is.null(split.by) && !split.by %in% colnames(colData(x))) {
+            stop("The 'split.by' column does not exist in colData(x).", 
+                 call. = FALSE)
+        }
+
+        ############################## Calculate metrics #######################
+        # Split the data by subject (or other grouping variable) if required
+        if (!is.null(split.by)) {
+            split_data <- split(x, f = colData(x)[[split.by]])
+        } else {
+            split_data <- list(x)  
+        }
+
+        # Calculate metrics for each split subject/group
+        results_list <- lapply(split_data, function(sub_data) {
+            # Initialize a list to hold results for each metric
+            metric_results <- list()
+            
+            # Loop through each metric and calculate it
+            for (metric in metrics) {
+                # Check if the metric is supported and calculate accordingly
+                if (metric == "CRT") {
+                    crt_results <- .calculateCRT(sub_data, thresholds[["CRT"]])
+                    metric_results[["CRT"]] <- crt_results
+                } else {
+                    stop(paste("Metric", metric, "is not supported yet."), 
+                         call. = FALSE)
+                }
+            }
+
+            # Add the calculated metrics to the metadata (colData or rowData)
+            for (metric in names(metric_results)) {
+                colData(sub_data)[[metric]] <- metric_results[[metric]]
+            }
+            
+            return(sub_data)
+        })
```

Comment:
It would be more efficient to:

0. Get matrix
1. Get the list which has vectors as elements. Each vector represent members of each group
2. Loop through this list
3. Select the group members from the matrix
4. Calculate values
5. Combine all the results and sort values based on sample names
6. Add to colData (in add* method)

Reply:
I am not convinced the right location should be colData as the interest is taxa related.

Reply:
True, I followed the code and did not think about it. As it these features are global (for whole dataset), colData is not correct place indeed

I think the output could be similar to getPrevalent and getRare, i.e., vector

Reply:
I think one vector could be calculated for whole dataset for now. As you mentioned, the multi-group thing is more general question, and is affecting also the mia functions

Reply:
Yes - note that this can only be calculated in the case where the whole data set has some continuous variable that extends all samples (like is the case when there is time dimension for a single subject). If we have multiple groups then this is not defined for the whole data set.

Reply:
I might have missed something, but I cannot see direct linkage of time series to this concept of conditionally/persistent rare/abundant taxa. It seems that this is more general concept?

Based on the following article, I summarized the concept

Sizhong Yang, Matthias Winkel, Dirk Wagner, Susanne Liebner, Community structure of rare methanogenic archaea: insight from a single functional group, FEMS Microbiology Ecology, Volume 93, Issue 11, November 2017, fix126, https://doi.org/10.1093/femsec/fix126

Rare taxa:

- Average relative abundance below 1% accross all samples

Abundant taxa:

- Average relative abundance above 1% accross all samples

Conditionally rare taxa:

> microbial taxa that are typically in low abundance in one locality but occasionally become prevalent in at least one other sample

- CRT was assigned if an OTU exhibits a maximum relative abundance at least 100 times higher than its minimum value (max:min > 100)

Permanently rare taxa:

- If the (max/min) ratio is less than 5, that is the abundance rarely fluctuates among samples, it will be grouped into the permanently rare taxa (PER)

Rare and abundant taxa can be already calculated by mia functions:

`getRare(tse, assay.type = "relabundance", detection = 1/100)`
`getPrevalent(tse, assay.type = "relabundance", detection = 1/100)`


I see that these conditionally rare and permanently rare taxa are very close to `getRare()` and `getPrevalent()`. As they do not necessarily deal with time series, should these methods be in mia instead: `getConditionallyRare()` & `getPermanentlyRare()`

Reply:
Yep, I mixed it up, no time series needed. 

CRT seems like a bit strange concept in the end. I am not sure if there should be much difference whether the minimum count is 1, 2, or 3 reads but this can make a huge difference in the max:min calculation. But yes, it might be relevant to recognize taxa that have vast variation. Variance would not capture all these cases, if the blooms cover only few samples. Shouldn't this be called conditionally abundant, rather than conditionally rare..?

PER: is the rarity also evaluated, or just max/min ratio? some taxa might be abundant and prevalent, while having a low max/min ratio. At least in principle. Then it would be misleading to call them PER. The interesting PER are taxa that might be prevalent but always stay under a certain threshold and never bloom.

We could add the methods as suggested and it could go to mia. If this could be made into a more generic index function (like addAlpha for samples), that could help to simplify maintenance and adding possible other measures later. Would that be feasible..?

On the other if this seem time-consuming to finalize, it could be left for later.